### PR TITLE
[WIP] vim-patch:7.4.1494

### PR DESCRIPTION
**Pull request by prollings**
Problem:    clr_history() does not work properly.
Solution:   Increment hisptr. Add a test. (Yegappan Lakshmanan)

https://github.com/vim/vim/commit/119d4693e06e68d4f099aa7287e375ae3d265fd0

Note: Only added a test. The change to `histptr` had already been made.

### CODE REVIEW

**Comment by prollings** [comment]
https://github.com/neovim/neovim/pull/5049/commits/ace84c9da8acbfca962ed962de0e6023a8a4c090#diff-d60a5538392b657181a7ad132099867aR53

Test for `''` Fails. But testing manually in nvim, it appears to be fine.

File: src/nvim/testdir/Makefile [hunk 1]
```diff
@@ -33,16 +33,17 @@ SCRIPTS := \
...
+      test_hardcopy.res \
+      test_help_tagjump.res \
+      test_history.res \
+      test_langmap.res \
+      test_menu.res \
+      test_syntax.res \
+      test_timers.res \
+      test_unlet.res \
+      test_viml.res \
+      test_alot.res
```
Comment (justinmk) [inline src/nvim/testdir/Makefile]: make requires tabs. is this indented with tab?

**Comment by justinmk** [comment]
Changed to WIP until we figure out what's wrong.

**Comment by jamessan** [comment]
> The change to histptr had already been made.

It has, but we're calling [hist_free_entry] in the loop, while Vim is calling [clear_hist_entry].  That difference may be relevant.

[hist_free_entry]: https://github.com/neovim/neovim/blob/master/src/nvim/ex_getln.c#L4804
[clear_hist_entry]: https://github.com/vim/vim/blob/119d4693e06e68d4f099aa7287e375ae3d265fd0/src/ex_getln.c#L5785

**Comment by jamessan** [comment]
Indeed.  Applying this

```diff
diff --git a/src/nvim/ex_getln.c b/src/nvim/ex_getln.c
index 1a97dc3..7f63541 100644
--- a/src/nvim/ex_getln.c
+++ b/src/nvim/ex_getln.c
@@ -4801,7 +4801,7 @@ int clr_history(const int histype)
   if (hislen != 0 && histype >= 0 && histype < HIST_COUNT) {
     histentry_T *hisptr = history[histype];
     for (int i = hislen; i--; hisptr++) {
-      hist_free_entry(hisptr);
+      clear_hist_entry(hisptr);
     }
     hisidx[histype] = -1;  // mark history as cleared
     hisnum[histype] = 0;   // reset identifier counter
```

makes all tests (oldtest, functionaltest, and unittest) pass.  The change (from clear_hist_entry -> hist_free_entry) was introduced in 244dbe3a77bf548f73d8781da7327f30e818b08a by @ZyX-I.

**Comment by justinmk** [comment]
So probably https://github.com/neovim/neovim/pull/5359 is not needed. Though it's uncomfortable that all of this is so fragile.

**Comment by ZyX-I** [comment]
@jamessan In `clr_history()` vim does not use *just* clear_hist_entry, it uses vim_free+clear_hist_entry. hist_free_entry is a thing which replaced this very common pattern: call vim_free(hisptr->hisstr), then call clear_hist_entry(hisptr), so your patch must be leaking memory. If all tests pass without triggering memory sanitizer error then they are incomplete. Additional reason was that to free history entry after ShaDa was introduced one must not only call vim_free (xfree) on hisstr, but also deref additional_elements.

**Comment by ZyX-I** [comment]
@prollings Which test fails? Link is broken.

**Comment by jamessan** [comment]
After rebasing this PR on current master, these are the results of the added test:

```
From test_history.vim:
Executing Test_History()
Executed 1 test
1 FAILED
Found errors in Test_History():
function RunTheTest[9]..Test_History[2]..History_Tests line 3: Expected True but got 0
function RunTheTest[9]..Test_History[2]..History_Tests line 7: Expected True but got 0
function RunTheTest[9]..Test_History[2]..History_Tests line 8: Expected True but got 0
function RunTheTest[9]..Test_History[2]..History_Tests line 9: Expected 'buffers' but got ''
function RunTheTest[9]..Test_History[2]..History_Tests line 10: Expected 'ls' but got ''
function RunTheTest[9]..Test_History[2]..History_Tests line 11: Expected 'ls' but got ''
function RunTheTest[9]..Test_History[2]..History_Tests line 14: Expected 2 but got -1
function RunTheTest[9]..Test_History[2]..History_Tests line 15: Expected True but got 0
function RunTheTest[9]..Test_History[2]..History_Tests line 17: Expected 1 but got -1
function RunTheTest[9]..Test_History[2]..History_Tests line 18: Expected 'ls' but got ''
function RunTheTest[9]..Test_History[2]..History_Tests line 20: Expected True but got 0
function RunTheTest[9]..Test_History[2]..History_Tests line 21: Expected True but got 0
function RunTheTest[9]..Test_History[2]..History_Tests line 22: Expected 'ls' but got ''
function RunTheTest[9]..Test_History[2]..History_Tests line 23: Expected 4 but got -1
function RunTheTest[9]..Test_History[2]..History_Tests line 29: Expected True but got 0
function RunTheTest[9]..Test_History[2]..History_Tests line 30: Expected 'ls' but got ''
```

**Comment by jamessan** [comment]
The failing call to `History_Tests` is when `a:hist` is the empty string.  I'm not sure why that's being tested, since that's not a valid history name according to `:help hist-names`.

**Comment by jamessan** [comment]
Yeah, applying this does make the tests pass.

```diff
diff --git a/src/nvim/testdir/test_history.vim b/src/nvim/testdir/test_history.vim
index ee6acff..bdb2027 100644
--- a/src/nvim/testdir/test_history.vim
+++ b/src/nvim/testdir/test_history.vim
@@ -50,7 +51,7 @@ function History_Tests(hist)
 endfunction
 
 function Test_History()
-  for h in ['cmd', ':', '', 'search', '/', '?', 'expr', '=', 'input', '@', 'debug', '>']
+  for h in ['cmd', ':', 'search', '/', '?', 'expr', '=', 'input', '@', 'debug', '>']
     call History_Tests(h)
   endfor
 
```

I'll raise an issue on Vim about the `''` test, since that shouldn't be there.

**Comment by jamessan** [comment]
Bah, no, this is documented in Vim's source (but not the docs).

```c
get_histtype(char_u *name)
{
    int		i;
    int		len = (int)STRLEN(name);

    /* No argument: use current history. */
    if (len == 0)
	return hist_char2type(ccline.cmdfirstc);
```

So, the Vim needs to fix its documentation and we need to fix our regression against Vim's behavior.

**Comment by Shougo** [comment]
@prollings Can works for this?
You are no github activities in some months.

I think other people have to fix the PR.

**Comment by ZyX-I** [comment]
@jamessan Or remove `if(len==0)` check, in any case that can go this way only through `hist*()` functions for which this is not documented. Before keeping this feature it is better ask what for and where this “current history” may be and is used.

**Comment by ZyX-I** [comment]
Note that if `ccline.cmdfirstc` is NUL, Vim will use search history and not command like in #5359. And there is no function which exposes what history will be used, so there is even no simple (i.e. different from “modify history and find what history was modified”) way to check what result this feature will yield.

**Comment by jamessan** [comment]
It seems like the intention for supporting an empty string in the `hist*` functions is similar to having `getcmdtype()`.  It's really only useful when evaluating an expression in a mapping so you don't have to know what the current history/cmd type.

Most likely, using `hist*(getcmdtype(), ...)` is a better replacement for `hist*('', ...)` since it's both using a documented interface and more explicit.

So, I guess the decision is whether to support Vim's undocumented (except in source) `''` hist-name or remove that part of the test.

**Comment by justinmk** [comment]
So `get_cmdline_type()` can be used when NUL is noticed

**Comment by jamessan** [comment]
I sent a mail to [vim-dev], since it seems like this functionality is just an afterthought.

[vim-dev]: https://www.mail-archive.com/vim_dev@googlegroups.com/msg42387.html

**Comment by jamessan** [comment]
> I sent a mail to vim-dev, since it seems like this functionality is just an afterthought.

No response on list, but Bram updated the documentation to specify that an [empty string is a valid](https://github.com/vim/vim/blob/3e496b0ea31996b665824f45664dee1fdd73c4d0/runtime/doc/eval.txt#L4791) hist-name.

**Comment by jamessan** [comment]
Continuing in #5503.

**Pull request closed without merging**